nv-codec-headers: adopt and refactor

AndersonTorres opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented:
Oh no, mass rebuild!

AndersonTorres commented:
@infinisil it looks like a bug to me! Can you verify?

I am using an external file to parameterize the `nv-codec-headers-${version}`.

infinisil commented:
Not a bug I'm afraid! To CI, the `nv-codec-headers-9` looks just like a new package defined with `pkgs.callPackage`.

Since `pkgs/by-name` can't yet handle package sets, I recommend refactoring this to:

```nix
  inherit (callPackages ../development/libraries/nv-codec-headers { })
    nv-codec-headers-9 nv-codec-headers-10;
```

or if you want to also expose the set of versions:

```nix
  nv-codec-headers-versions = callPackages ../development/libraries/nv-codec-headers { };
  inherit (nv-codec-headers-versions)
    nv-codec-headers-9 nv-codec-headers-10;
```

See also https://github.com/NixOS/nixpkgs/pull/285922#issuecomment-1925940231 or https://github.com/NixOS/nixpkgs/pull/287081 for some discussion on this.

AndersonTorres committed:
nv-codec-headers-versions: init

It globs all versions of nv-codec-headers kept/needed on Nixpkgs.

The default is kept at 9, as before.

AndersonTorres commented:
`callPackages` in plural? Oh jeez

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3486

SuperSandro2000 commented on pkgs/development/libraries/nv-codec-headers/make-nv-codec-headers.nix:
```diff
@@ -0,0 +1,31 @@
+
+  meta = {
+    description = "FFmpeg version of headers for NVENC";
+    homepage = "https://ffmpeg.org/";
```
```suggestion
    homepage = "https://ffmpeg.org/";
    downloagPage = "https://git.videolan.org/?p=ffmpeg/nv-codec-headers.git";
```
because when you paste the git url in a browser, you just get a 404

SuperSandro2000 commented on pkgs/development/libraries/nv-codec-headers/default.nix:
```diff
@@ -1,22 +1,28 @@
-    rev = "n${version}";
-    sha256 = "1xfvb3mhz6wfx9c732888xa82ivaig903lhvvrqqzs31qfznsplh";
+let
+  make-nv-codec-headers = (import ./make-nv-codec-headers.nix) {
```
Shouldn't this file be called generic.nix or common.nix?

AndersonTorres replied:
_Data maxima venia_, such name for a file is ridiculous and convey no serious, palpable meaning.
Something like `make-XYZ` makes clear that this file contains a function (or something like that) that makes `XYZ`.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -24075,10 +24075,16 @@ with pkgs;
-  nv-codec-headers-10 = callPackage ../development/libraries/nv-codec-headers/10_x.nix { };
-  nv-codec-headers-11 = callPackage ../development/libraries/nv-codec-headers/11_x.nix { };
-  nv-codec-headers-12 = callPackage ../development/libraries/nv-codec-headers/12_x.nix { };
+  # splicing magic
+  nv-codec-headers-versions = callPackages ../development/libraries/nv-codec-headers { };
+  inherit (nv-codec-headers-versions)
+    nv-codec-headers-9
+    nv-codec-headers-10
+    nv-codec-headers-11
+    nv-codec-headers-12
+  ;
+  # A default nv-codec-headers to make people happy
+  nv-codec-headers = nv-codec-headers-versions.nv-codec-headers-9;
```
```suggestion
  inherit (callPackages ../development/libraries/nv-codec-headers { })
    nv-codec-headers-9
    nv-codec-headers-10
    nv-codec-headers-11
    nv-codec-headers-12
  ;
  nv-codec-headers = nv-codec-headers-9;
```
I don't think we need that new group here
